pkg/policy/rule.go

```diff
@@ -16,55 +16,114 @@ package policy
 
 import (
 	"fmt"
-	"strings"
 
-	"github.com/cilium/cilium/common"
 	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy/api"
 )
 
-type PolicyRule interface {
-	// Resolve must resolve any internal label members to the full path
-	// assuming that the rule is attached to the specified node.
-	Resolve(node *Node) error
+type rule struct {
+	api.Rule
+}
+
+func (r *rule) String() string {
+	return fmt.Sprintf("%v", r.EndpointSelector)
+}
+
+func (r *rule) validate() error {
+	if r == nil {
+		return fmt.Errorf("nil rule")
+	}
+
+	if len(r.EndpointSelector) == 0 {
+		return fmt.Errorf("empty EndpointSelector")
+	}
+
+	return nil
+}
+
+func mergeL4Port(r api.PortRule, p api.PortProtocol, proto string, resMap L4PolicyMap) int {
+	fmt := fmt.Sprintf("%s:%d", proto, p.Port)
```

Comment:
Either is fine, it doesn't matter. Never exposed. Will change this if you look this better.